common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/BlockStoreClient.java

```diff
@@ -188,6 +212,31 @@ public void onFailure(Throwable t) {
     }
   }
 
+  private <T> void retry(
+      int times,
+      final int maxRetries,
+      int delayMs,
+      Supplier<CompletableFuture<T>> action,
+      CompletableFuture<T> future) {
+    action.get()
+            .thenAccept(future::complete)
+            .exceptionally(e -> {
+              boolean isIOException = e instanceof IOException
+                      || (e.getCause() != null && e.getCause() instanceof IOException);
```

Comment:
This should also include Sasl timeout (see [here](https://github.com/apache/spark/blob/692ec99bc7884cc998afbf63e4ef53053c0c9dd7/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RetryingBlockTransferor.java#L218)). In future, if we want to add more exceptions for which there will be a retry, then we have to be careful to update all the places where this logic is use, so I think we should come up a with a way that this is not duplicated.

Reply:
Indeed, the retry logic of enable sasl retries is missing here.